spices/SPICE-0004-HTTP-Proxy-Support.adoc

```diff
@@ -0,0 +1,180 @@
+// suppress inspection "HttpUrlsUsage" for whole file
+= HTTP Proxy Support
+
+* Proposal: link:./SPICE-0004-HTTP-Proxy-Support.adoc[SPICE-0004]
+* Author: link:https://github.com/bioball[Dan Chao]
+* Status: TBD
+* Implemented in: TBD
+* Category: Language, Standard Library
+
+== Introduction
+
+This SPICE describes support for HTTP proxying.
+
+== Motivation
+
+Some environments do not allow direct connections to the internet, instead requiring that connections go through an intermediary proxy.
+
+To enable Pkl in these situations, this SPICE provides a way to redirect outbound requests through said proxies.
+
+== Proposed Solution
+
+Pkl will support HTTP proxies.
+Proxy settings will be added, and this will control how Pkl makes HTTP requests.
+
+== Detailed design
+
+=== Configuring proxies
+
+A proxy can be set via the following, in order of precedence (least to greatest):
+
+1. `~/.pkl/settings` file
+2. `PklProject` file (in `evaluatorSettings`)
+3. CLI Flags/pkl-gradle options/`HttpClient#Builder`/server mode message passing API
+
+==== Environment Variables
+
+Proxies will _not_ be configured via environment variables.
```

Comment:
I think this is a mistake. Env vars are *the* way to configure proxy settings for command-line tools. Unlike the listed hypothetical worries, having to configure each tool individually is a very real problem.

Reply:
I'm open to the argument, but: another concern here is: how does this play with `--env-var`? What happens if you do:

```
pkl eval --env-var http_proxy=http://some.proxy foo.pkl
```

It feels like this shouldn't affect proxy settings, because this is ostensibly just a way to configure what `read("env:")` will return.

But if that doesn't affect the proxy, this is possibly a strange behavior:

```
# talk via the proxy
export http_proxy=http://some.proxy pkl eval foo.pkl

# does not talk via the proxy
pkl eval --env-var http_proxy=http://some.proxy foo.pkl
```

There's also plenty of languages/tools whose proxies aren't configured by env vars. Just some quick skimming (as far as I can tell):

* Node.js
* Java
* Rust, Cargo
* PHP

Reply:
Java supports OS proxy settings, and there are signs that env vars will be reconsidered. Supporting neither OS settings nor env vars seems problematic, forcing users to script a solution specifically for Pkl.

Reply:
Not sure I'm sold on configuring proxies via env vars. Nothing else in Pkl in configured using them, also, It may lead to "works on my machine" problems where you have the right env vars set.

Reply:
Supporting OS proxy settings sounds more enticing than supporting envvar-based configuration, if that's possible.